src/ids.rs

```diff
@@ -0,0 +1,109 @@
+/// Maps u64 integers to a consecutive range of ids
+#[derive(Debug)]
+pub struct IdTable {
+    // map u64 id x to u32 by storing a sorted mapping table for each value of x / 2^32
+    data: Vec<Vec<(u32, u32)>>,
+    num_ids: u32,
+}
+
+#[derive(Debug)]
+pub struct IdTableBuilder {
+    // stored the same data as IdTable, but not yet sorted
+    data: Vec<Vec<(u32, u32)>>,
+    next_id: u32,
+}
+
+impl IdTableBuilder {
+    pub fn new() -> Self {
+        Self {
+            data: Vec::new(),
+            next_id: 0,
+        }
+    }
+
+    /// Inserts an Id and returns a mapped index
+    pub fn insert(&mut self, x: u64) -> u32 {
+        let id_set = (x >> 32) as usize;
+        if self.data.len() <= id_set {
+            self.data.resize(id_set + 1, Vec::new());
+        }
+        self.data[id_set].push((x as u32, self.next_id));
+        let result = self.next_id;
+        self.next_id += 1;
+        result
+    }
+
+    /// Skips a few ids (e.g. to reserve them for other uses)
+    pub fn skip(&mut self, count: u32) {
+        self.next_id += count;
+    }
+
+    pub fn finalize(mut self) -> IdTable {
+        for mut x in &mut self.data {
+            x.sort();
+        }
+
+        IdTable {
+            data: self.data,
+            num_ids: self.next_id,
+        }
+    }
+}
+
+impl IdTable {
+    pub fn find(&self, x: u64) -> Option<u32> {
```

Comment:
Should not this be `get` like in `BTreeMap` or `HashMap`? Then you even don't have to adjust the usage of it (except for removal of &'s).